Update examples (1) (#7988)
ethereum/solidity: docs/examples/micropayment.rst

Recovering the Message Signer in Solidity
-----------------------------------------

In general, ECDSA signatures consist of two parameters,
``r`` and ``s``. Signatures in Ethereum include a third
parameter called ``v``, that you can use to verify which
account's private key was used to sign the message, and
the transaction's sender. Solidity provides a built-in
function `ecrecover <mathematical-and-cryptographic-functions>`_ that
accepts a message along with the ``r``, ``s`` and ``v`` parameters
and returns the address that was used to sign the message.

Extracting the Signature Parameters
-----------------------------------

Signatures produced by web3.js are the concatenation of ``r``,
``s`` and ``v``, so the first step is to split these parameters
apart. You can do this on the client-side, but doing it inside
the smart contract means you only need to send one signature
parameter rather than three. Splitting apart a byte array into

> component parts is a mess, so we use

Reviewer: ```suggestion
its constituent parts is a mess, so we use
```

Writing a Simple Payment Channel
================================

Alice now builds a simple but complete implementation of a payment
channel. Payment channels use cryptographic signatures to make
repeated transfers of Ether securely, instantaneously, and without transaction fees.

What is a Payment Channel?
--------------------------

Payment channels allow participants to make repeated transfers of Ether
without using transactions. This means that you can avoid the delays and
fees associated with transactions. We are going to explore a simple
unidirectional payment channel between two parties (Alice and Bob). It involves three steps:

    1. Alice funds a smart contract with Ether. This "opens" the payment channel.
    2. Alice signs messages that specify how much of that Ether is owed to the recipient. This step is repeated for each payment.
    3. Bob "closes" the payment channel, withdrawing their portion of the Ether and sending the remainder back to the sender.

.. note::

>   Only steps 1 and 3 require Ethereum transactions, step 2 means that the sender

Reviewer: Is the extra indentation in this and the next few lines intended?

Author: Should be the same indentation as the non-broken line was before.

Reviewer: The second line onwards are left aligned in the original text but not in this PR, is that also intended?

  transmits a cryptographically signed message to the recipient via off chain
  methods (e.g. email). This means only two transactions are required to support
  any number of transfers.

Bob is guaranteed to receive their funds because the smart contract escrows the
Ether and honours a valid signed message. The smart contract also enforces a
timeout, so Alice is guaranteed to eventually recover their funds even if the
recipient refuses to close the channel. It is up to the participants in a payment
channel to decide how long to keep it open. For a short-lived transaction,

> such as paying an internet café for each minute of network access, or for a

Reviewer: I would suggest rephrasing this sentence as follows or something like that: `such as paying an internet café for each minute of network access, the payment channel may be kept open for a limited duration. On the other hand, for a recurring payment, such as paying an employee an hourly wage, the payment could be kept open for several months or years.`

Author: I just added line breaks. Can you make a PR after this is merged?

Reviewer: Okay (CC https://github.com/ethereum/solidity/issues/7990)

> longer relationship, such as paying an employee an hourly wage, a payment 

Reviewer: ```suggestion
longer relationship, such as paying an employee an hourly wage, a payment
```

> ould last for months or years.

Reviewer: ```suggestion
could last for months or years.
```

each message specifies a cumulative total amount of Ether owed, rather than the
amount of the individual micropayment. The recipient will naturally choose to
redeem the most recent message because that is the one with the highest total.
The nonce per-message is not needed anymore, because the smart contract only

> honors a single message. The address of the smart contract is still used

Reviewer: ```suggestion
honours a single message. The address of the smart contract is still used
```
